### PR TITLE
Adds Toyota Highlander (+Hybrid) support

### PR DESCRIPTION
**Pull request by daehahn**
Working on 2017 Highlander and 2018 Highlander Hybrid. It's basically working, but definitely need a lot of  turning.

### CODE REVIEW

File: selfdrive/car/toyota/values.py [hunk 1]
```diff
@@ -34,33 +36,41 @@ class ECU:
...
-               (0x32E, ECU.APGS, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4, CAR.COROLLA), 0,  20, '\x00\x00\x00\x00\x00\x00\x00\x00'),
-               (0x33E, ECU.DSU, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH), 0,  20, '\x0f\xff\x26\x40\x00\x1f\x00'),
-               (0x365, ECU.DSU, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH), 0,  20, '\x00\x00\x00\x80\x03\x00\x08'),
-               (0x365, ECU.DSU, (CAR.RAV4, CAR.COROLLA), 0,  20, '\x00\x00\x00\x80\xfc\x00\x08'),
-               (0x366, ECU.DSU, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH), 0,  20, '\x00\x00\x4d\x82\x40\x02\x00'),
+               (0X161, ECU.DSU, (CAR.HIGHLANDERH, CAR.HIGHLANDER), 1,  7, '\x00\x1e\x00\xd4\x00\x00\x5b'),
+
+               (0x32E, ECU.APGS, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4, CAR.COROLLA, CAR.HIGHLANDERH, CAR.HIGHLANDER), 0,  20, '\x00\x00\x00\x00\x00\x00\x00\x00'),
+               (0x33E, ECU.DSU, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.HIGHLANDERH), 0,  20, '\x0f\xff\x26\x40\x00\x1f\x00'),
+               (0x343, ECU.DSU, (CAR.HIGHLANDERH, CAR.HIGHLANDER), 0,  20, '\x00\x00\x83\x00\x00\x00\x00\xd1'),
```
Comment (rbiasini) [inline selfdrive/car/toyota/values.py]: @daehahn you are sending 0x343 twice. This is the ACC_CONTROL msg

Comment (daehahn) [reply selfdrive/car/toyota/values.py]: I see. That is the reason why ACCEL_CMD is not working well. I will fix it. THX!!

File: selfdrive/car/toyota/interface.py [hunk 2]
```diff
@@ -233,8 +242,8 @@ def update(self, c):
     ret.cruiseState.speed = self.CS.v_cruise_pcm * CV.KPH_TO_MS
     ret.cruiseState.available = bool(self.CS.main_on)
     ret.cruiseState.speedOffset = 0.
-    if self.CP.carFingerprint == CAR.RAV4H:
-      # ignore standstill in hybrid rav4, since pcm allows to restart without
+    if (self.CP.carFingerprint == CAR.RAV4H) or (self.CP.carFingerprint == CAR.HIGHLANDERH):
```
Comment (rbiasini) [inline selfdrive/car/toyota/interface.py]: `if self.CP.carFingerprint in [CAR.RAV4H, CAR.HIGHLANDERH]:`

Comment (daehahn) [reply selfdrive/car/toyota/interface.py]: Finished touch up!

File: README.md [hunk 3]
```diff
@@ -80,7 +80,8 @@ Supported Cars
 | Toyota        | Rav4 2018                 | All                  | Yes     | Yes<sup>2</sup>| 20mph            | 0mph           |
 | Toyota        | Rav4 Hybrid 2017          | All                  | Yes     | Yes<sup>2</sup>| 0mph             | 0mph           |
 | Toyota        | Rav4 Hybrid 2018          | All                  | Yes     | Yes<sup>2</sup>| 0mph             | 0mph           |
-
+| Toyota        | Highlander 2017           | All                  | Yes     | Yes<sup>2</sup>| 20mph            | 0mph           |
+| Toyota        | Highlander Hybrid 2018    | All                  | Yes     | Yes<sup>2</sup>| 0mph             | 0mph           |
```
[the remark this answers is not in the document]
Comment (daehahn) [reply README.md]: I did it.

File: selfdrive/car/toyota/values.py [hunk 4]
```diff
@@ -34,33 +36,40 @@ class ECU:
...
-               (0x32E, ECU.APGS, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4, CAR.COROLLA), 0,  20, '\x00\x00\x00\x00\x00\x00\x00\x00'),
-               (0x33E, ECU.DSU, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH), 0,  20, '\x0f\xff\x26\x40\x00\x1f\x00'),
-               (0x365, ECU.DSU, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH), 0,  20, '\x00\x00\x00\x80\x03\x00\x08'),
-               (0x365, ECU.DSU, (CAR.RAV4, CAR.COROLLA), 0,  20, '\x00\x00\x00\x80\xfc\x00\x08'),
-               (0x366, ECU.DSU, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH), 0,  20, '\x00\x00\x4d\x82\x40\x02\x00'),
+               (0X161, ECU.DSU, (CAR.HIGHLANDERH, CAR.HIGHLANDER), 1,  7, '\x00\x1e\x00\xd4\x00\x00\x5b'),
+
+               (0x32E, ECU.APGS, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4, CAR.COROLLA, CAR.HIGHLANDERH, CAR.HIGHLANDER), 0,  20, '\x00\x00\x00\x00\x00\x00\x00\x00'),
+               (0x33E, ECU.DSU, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.HIGHLANDERH), 0,  20, '\x0f\xff\x26\x40\x00\x1f\x00'),
+               (0x344, ECU.DSU, (CAR.HIGHLANDERH, CAR.HIGHLANDER), 0,  20, '\x00\x00\x01\x00\x00\x00\x00\x50'),
```
Comment (rbiasini) [inline selfdrive/car/toyota/values.py]: this is the same as the 344 defined above, except for the frequency. You sure about it?

Comment (daehahn) [reply selfdrive/car/toyota/values.py]: No I'm not sure not. That looks mistakenly place while migrated from 
https://github.com/ConservativeEconomist/openpilot/blob/vanillagorillaa-highlander-port/selfdrive/car/toyota/values.py

File: selfdrive/car/toyota/values.py [hunk 5]
```diff
@@ -34,33 +36,40 @@ class ECU:
...
+               (0x489, ECU.CAM, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4, CAR.COROLLA, CAR.HIGHLANDERH, CAR.HIGHLANDER), 0, 100, '\x00\x00\x00\x00\x00\x00\x00'),
+               (0x48a, ECU.CAM, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4, CAR.COROLLA, CAR.HIGHLANDERH, CAR.HIGHLANDER), 0, 100, '\x00\x00\x00\x00\x00\x00\x00'),
+               (0x48b, ECU.CAM, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4, CAR.COROLLA, CAR.HIGHLANDERH, CAR.HIGHLANDER), 0, 100, '\x66\x06\x08\x0a\x02\x00\x00\x00'),
+               (0x4d3, ECU.CAM, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4, CAR.COROLLA, CAR.HIGHLANDERH, CAR.HIGHLANDER), 0, 100, '\x1C\x00\x00\x01\x00\x00\x00\x00'),
                (0x130, ECU.CAM, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4, CAR.COROLLA), 1, 100, '\x00\x00\x00\x00\x00\x00\x38'),
-               (0x466, ECU.CAM, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4), 1, 100, '\x20\x20\xAD'),
+               (0x466, ECU.CAM, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4, CAR.HIGHLANDERH, ), 1, 100, '\x20\x20\xAD'),
```
Comment (rbiasini) [inline selfdrive/car/toyota/values.py]: you sure the highlander non-hybrid does not have 0x466?

Comment (daehahn) [reply selfdrive/car/toyota/values.py]: We (me and Phil Lee) need to fine tuning those STATIC_MSGS. Let us guide. That this is CAN message map build by myself for these tuning. https://drive.google.com/drive/u/0/folders/1u_K-uB0PXv0OeZ5iUNoRuBdUMMt61LBb

File: selfdrive/car/toyota/values.py (hunk 4 above)
Comment (rbiasini) [reply selfdrive/car/toyota/values.py]: yeah, just verified. This msg is at 20Hz, so it needs to be sent every 5 steps (each step is 0.01s).
So CAR.HIGHLANDERH, CAR.HIGHLANDER should just be added to line 37

File: selfdrive/car/toyota/values.py [hunk 6]
```diff
@@ -34,33 +36,40 @@ class ECU:
...
                (0x367, ECU.CAM, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4, CAR.COROLLA), 0,  40, '\x06\x00'),
-
-               (0x414, ECU.CAM, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4, CAR.COROLLA), 0, 100, '\x00\x00\x00\x00\x00\x00\x17\x00'),
-               (0x489, ECU.CAM, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4, CAR.COROLLA), 0, 100, '\x00\x00\x00\x00\x00\x00\x00'),
-               (0x48a, ECU.CAM, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4, CAR.COROLLA), 0, 100, '\x00\x00\x00\x00\x00\x00\x00'),
-               (0x48b, ECU.CAM, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4, CAR.COROLLA), 0, 100, '\x66\x06\x08\x0a\x02\x00\x00\x00'),
-               (0x4d3, ECU.CAM, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4, CAR.COROLLA), 0, 100, '\x1C\x00\x00\x01\x00\x00\x00\x00'),
+               (0x367, ECU.CAM, (CAR.HIGHLANDERH, CAR.HIGHLANDER), 0,  40, '\x00\x00'),
+
+               (0x411, ECU.DSU, (CAR.HIGHLANDERH, CAR.HIGHLANDER), 0,  100, '\x00\x20\x00\x00\x10\x00\x80\x00'),
```
Comment (rbiasini) [inline selfdrive/car/toyota/values.py]: these are alerts already sent in carcontroller (like 0x343). This needs to be removed from here.

File: selfdrive/car/toyota/values.py [hunk 7]
```diff
@@ -34,33 +36,40 @@ class ECU:
...
+               (0x412, ECU.CAM, (CAR.HIGHLANDERH, CAR.HIGHLANDER), 0,  100, '\x68\x04\x0a\x00\x00\x2c\x38\x02'),
```
Comment (rbiasini) [inline selfdrive/car/toyota/values.py]: these are alerts already sent in carcontroller (like 0x343). This needs to be removed from here.

File: selfdrive/car/toyota/values.py [hunk 8]
```diff
@@ -34,33 +36,40 @@ class ECU:
...
+               (0x366, ECU.DSU, (CAR.HIGHLANDERH, CAR.HIGHLANDER), 0,  20, '\x00\xb1\x07\xff\x08\xff\x00'),
 
                (0x367, ECU.CAM, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4, CAR.COROLLA), 0,  40, '\x06\x00'),
-
-               (0x414, ECU.CAM, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4, CAR.COROLLA), 0, 100, '\x00\x00\x00\x00\x00\x00\x17\x00'),
-               (0x489, ECU.CAM, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4, CAR.COROLLA), 0, 100, '\x00\x00\x00\x00\x00\x00\x00'),
-               (0x48a, ECU.CAM, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4, CAR.COROLLA), 0, 100, '\x00\x00\x00\x00\x00\x00\x00'),
-               (0x48b, ECU.CAM, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4, CAR.COROLLA), 0, 100, '\x66\x06\x08\x0a\x02\x00\x00\x00'),
-               (0x4d3, ECU.CAM, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4, CAR.COROLLA), 0, 100, '\x1C\x00\x00\x01\x00\x00\x00\x00'),
+               (0x367, ECU.CAM, (CAR.HIGHLANDERH, CAR.HIGHLANDER), 0,  40, '\x00\x00'),
```
Comment (rbiasini) [inline selfdrive/car/toyota/values.py]: the value of this message varies. Sometimes it's set to '\x06\x00'... for simplicity let's remove this line and just add the added Highlander to the previous line.

File: selfdrive/car/toyota/values.py [hunk 9]
```diff
@@ -34,33 +36,40 @@ class ECU:
...
+               (0X161, ECU.DSU, (CAR.HIGHLANDERH, CAR.HIGHLANDER), 1,  7, '\x00\x1e\x00\xd4\x00\x00\x5b'),
+
+               (0x32E, ECU.APGS, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.RAV4, CAR.COROLLA, CAR.HIGHLANDERH, CAR.HIGHLANDER), 0,  20, '\x00\x00\x00\x00\x00\x00\x00\x00'),
+               (0x33E, ECU.DSU, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.HIGHLANDERH), 0,  20, '\x0f\xff\x26\x40\x00\x1f\x00'),
+               (0x344, ECU.DSU, (CAR.HIGHLANDERH, CAR.HIGHLANDER), 0,  20, '\x00\x00\x01\x00\x00\x00\x00\x50'),
+               (0x365, ECU.DSU, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.HIGHLANDERH), 0,  20, '\x00\x00\x00\x80\x03\x00\x08'),
+               (0x365, ECU.DSU, (CAR.RAV4, CAR.COROLLA, CAR.HIGHLANDER), 0,  20, '\x00\x00\x00\x80\xfc\x00\x08'),
+               (0x366, ECU.DSU, (CAR.PRIUS, CAR.RAV4H, CAR.LEXUS_RXH, CAR.HIGHLANDERH), 0,  20, '\x00\x00\x4d\x82\x40\x02\x00'),
                (0x366, ECU.DSU, (CAR.RAV4, CAR.COROLLA), 0,  20, '\x00\x72\x07\xff\x09\xfe\x00'),
+               (0x366, ECU.DSU, (CAR.HIGHLANDERH, CAR.HIGHLANDER), 0,  20, '\x00\xb1\x07\xff\x08\xff\x00'),
```
Comment (rbiasini) [inline selfdrive/car/toyota/values.py]: 0x366 is sent out twice for highlander hybrid

**Pull request closed without merging**